Malformed-XML error messages no longer have filename + location

The error message for bad XML is missing the filename, line, and column location that it used to have.

I suspect this is due to the changes in #1004 (the `XmlTextReader` we're using no longer knows the path of the file, because we're giving it a stream instead.

<details>
<summary>Expected and actual output</summary>

```XML
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003"
      DefaultTargets="t1">
  <Target Name="t1">
    <Message Text= [p2.t1]/>
  </Target>
  <Target Name="t2">
    <Message Text="[p2.t2]"/>
  </Target>
</Project>
```

Current:

```
S:\work\1244940>\msbuild\bin\Bootstrap\15.0\Bin\MSBuild.exe pE.proj
Microsoft (R) Build Engine version 15.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 10/31/2016 11:14:10 AM.
Project "S:\work\1244940\pE.proj" on node 1 (default targets).
MSBUILD : error MSB4025: The project file could not be loaded. '[' is an unexpected token. The expected token is '"' or '''. Line 4, position 26. [S:\work\1244940\pE.proj]
Done Building Project "S:\work\1244940\pE.proj" (default targets) -- FAILED.


Build FAILED.

"S:\work\1244940\pE.proj" (default target) (1) ->
  MSBUILD : error MSB4025: The project file could not be loaded. '[' is an unexpected token. The expected token is '"' or '''. Line 4, position 26. [S:\work\1244940\pE.proj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.23
```

MSBuild 14:

```
S:\work\1244940>MSBuild.exe pE.proj
Microsoft (R) Build Engine version 14.0.25420.1
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 10/31/2016 11:14:39 AM.
Project "S:\work\1244940\pE.proj" on node 1 (default targets).
S:\work\1244940\pE.proj(4,26): error MSB4025: The project file could not be loaded. '[' is an unexpected token. The expected token is '"' or '''. Line 4, position 26.
Done Building Project "S:\work\1244940\pE.proj" (default targets) -- FAILED.


Build FAILED.

"S:\work\1244940\pE.proj" (default target) (1) ->
  S:\work\1244940\pE.proj(4,26): error MSB4025: The project file could not be loaded. '[' is an unexpected token. The expected token is '"' or '''. Line 4, position 26.

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.02
```
</details>

--- a/src/Shared/BuildEventFileInfo.cs
+++ b/src/Shared/BuildEventFileInfo.cs
@@ -97,6 +97,16 @@ namespace Microsoft.Build.Shared
             _endColumn = 0;
         }
 
+        /// <summary>
+        /// Creates an instance of this class using the information in the given XmlException and file location.
+        /// </summary>
+        internal BuildEventFileInfo(string file, XmlException e) : this(e)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(file, nameof(file));
+
+            _file = file;
+        }
+
         #endregion
 
         #region Properties

--- a/src/XMakeBuildEngine/Construction/ProjectRootElement.cs
+++ b/src/XMakeBuildEngine/Construction/ProjectRootElement.cs
@@ -2059,7 +2059,7 @@ namespace Microsoft.Build.Construction
 
                     if (xmlException != null)
                     {
-                        fileInfo = new BuildEventFileInfo(xmlException);
+                        fileInfo = new BuildEventFileInfo(fullPath, xmlException);
                     }
                     else
                     {

--- a/src/XMakeBuildEngine/Definition/Toolset.cs
+++ b/src/XMakeBuildEngine/Definition/Toolset.cs
@@ -1046,7 +1046,7 @@ namespace Microsoft.Build.Evaluation
                 catch (XmlException e)
                 {
                     // handle XML errors in the default tasks file
-                    ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(false, new BuildEventFileInfo(e), taskFileError, e.Message);
+                    ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(false, new BuildEventFileInfo(defaultTasksFile, e), taskFileError, e.Message);
                 }
                 catch (Exception e) when (ExceptionHandling.IsIoRelatedException(e))
                 {

--- a/src/XMakeBuildEngine/UnitTests/InvalidProjectFileException_Tests.cs
+++ b/src/XMakeBuildEngine/UnitTests/InvalidProjectFileException_Tests.cs
@@ -82,5 +82,36 @@ namespace Microsoft.Build.UnitTests
                 File.Delete(file);
             }
         }
+
+        /// <summary>
+        /// Regression test for https://github.com/Microsoft/msbuild/issues/1286
+        /// </summary>
+        [Fact]
+        public void LogErrorShouldHavePathAndLocation()
+        {
+            string file = Path.GetTempPath() + Guid.NewGuid().ToString("N");
+
+            try
+            {
+                File.WriteAllText(file, ObjectModelHelpers.CleanupFileContents(@"
+                    <Project ToolsVersion='msbuilddefaulttoolsversion' xmlns='msbuildnamespace'>
+                        <Target Name=[invalid] />
+                    </Project>"));
+
+                var _ = ObjectModelHelpers.BuildTempProjectFileExpectFailure(file);
+
+                Assert.True(false, "Loading an invalid project should have thrown an InvalidProjectFileException.");
+            }
+            catch (InvalidProjectFileException e)
+            {
+                Assert.Equal(3, e.LineNumber);
+                Assert.Equal(38, e.ColumnNumber);
+                Assert.Equal(file, e.ProjectFile); // https://github.com/Microsoft/msbuild/issues/1286
+            }
+            finally
+            {
+                File.Delete(file);
+            }
+        }
     }
 }

